### PR TITLE
DataGrid - ColumnChooser - When moving column from group panel to header column is invisible + refactoring

### PR DESCRIPTION
**Pull request by Tucchhaa**

### CODE REVIEW

File: js/ui/grid_core/ui.grid_core.column_chooser.js [hunk 1]
```diff
@@ -206,16 +171,45 @@ const columnChooserMembers = {
...
+
+        this._popupContainer.$wrapper()
+            .toggleClass(this.addWidgetPrefix(COLUMN_CHOOSER_DRAG_CLASS), !isSelectMode)
+            .toggleClass(this.addWidgetPrefix(COLUMN_CHOOSER_SELECT_CLASS), isSelectMode);
+
+        this._popupContainer.$content().addClass(this.addWidgetPrefix(COLUMN_CHOOSER_LIST_CLASS));
+
+        if(isSelectMode && !this._columnsController.isBandColumnsUsed()) {
+            this._popupContainer.$content().addClass(this.addWidgetPrefix(COLUMN_CHOOSER_PLAIN_CLASS));
+        }
```
Comment (williamvinogradov) [inline js/ui/grid_core/ui.grid_core.column_chooser.js]: Minor comment.

It's not good practice to use function call inside the if's condition
Anyway we should know the result of the ```isBandColumnsUsed```
Let's move it to a separate variable on the top of this function:

```suggestion
    setPopupAttributes: function() {
        const isSelectMode = this.isSelectMode();
        const isBandColumnUsed = this._columnsController.isBandColumnsUsed();

        this._popupContainer.setAria({
            role: 'dialog',
            label: messageLocalization.format('dxDataGrid-columnChooserTitle')
        });

        this._popupContainer.$wrapper()
            .toggleClass(this.addWidgetPrefix(COLUMN_CHOOSER_DRAG_CLASS), !isSelectMode)
            .toggleClass(this.addWidgetPrefix(COLUMN_CHOOSER_SELECT_CLASS), isSelectMode);

        this._popupContainer.$content().addClass(this.addWidgetPrefix(COLUMN_CHOOSER_LIST_CLASS));

        if(isSelectMode && !isBandColumnUsed) {
            this._popupContainer.$content().addClass(this.addWidgetPrefix(COLUMN_CHOOSER_PLAIN_CLASS));
        }
```

File: js/ui/grid_core/ui.grid_core.column_chooser.js [hunk 2]
```diff
@@ -278,8 +246,27 @@ const columnChooserMembers = {
...
+
+            this._columnChooserList.on('contentReady', e => {
+                deferUpdate(function() {
+                    const scrollable = e.component.getScrollable();
+                    scrollable.scrollTo({ y: scrollTop });
+
+                    // @ts-expect-error
+                    that.renderCompleted.fire();
+                });
+            });
```
Comment (williamvinogradov) [inline js/ui/grid_core/ui.grid_core.column_chooser.js]: Let's refactor here a little more.
Move this code to a separate entity, e.g. class ```ColumnChooserListScrollSynchronizer```:

```typescript
/* eslint-disable @typescript-eslint/no-parameter-properties */
import { deferUpdate } from '@js/core/utils/common';
import { DeferredObj } from '@js/core/utils/deferred';
import dxTreeView, { OptionChangedEvent, ContentReadyEvent } from '@js/ui/tree_view';

const TREE_VIEW_EVENTS = {
  optionChanged: 'optionChanged',
  contentReady: 'contentReady',
};

export class ColumnChooserListScrollSynchronizer {
  private scrollTop = 0;

  constructor(
    private readonly treeView: dxTreeView,
    private readonly renderCompleteCallback: () => void,
  ) {
  }

  start(): void {
    this.treeView.on(TREE_VIEW_EVENTS.optionChanged, this.onTreeViewOptionChanged);
    this.treeView.on(TREE_VIEW_EVENTS.contentReady, this.onTreeViewContentReady);
  }

  stop(): void {
    this.treeView.off(TREE_VIEW_EVENTS.optionChanged, this.onTreeViewOptionChanged);
    this.treeView.off(TREE_VIEW_EVENTS.contentReady, this.onTreeViewContentReady);
  }

  private onTreeViewOptionChanged(
    { component }: OptionChangedEvent,
  ): void {
    const scrollable = component.getScrollable();
    this.scrollTop = scrollable.scrollTop();
  }

  private onTreeViewContentReady(
    { component }: ContentReadyEvent,
  ): void | Promise<void> | DeferredObj<void> {
    return deferUpdate(() => {
      const scrollable = component.getScrollable();
      scrollable.scrollTo({ y: this.scrollTop });
      this.renderCompleteCallback?.();
    });
  }
}
```

And use it here, something like:
```suggestion
            this.columnChooserScrollSynchronizer?.stop();
            this.columnChooserScrollSynchronizer = new ColumnChooserListScrollSynchronizer(
              this._columnChooserList,
              () => { this.renderCompleted.fire(); }
            );
            this.columnChooserScrollSynchronizer.start();
```

File: js/ui/grid_core/ui.grid_core.column_chooser.js [hunk 3]
```diff
@@ -370,24 +359,46 @@ const columnChooserMembers = {
...
+    _updateItemSelection: function(columnIndex) {
+        const isRecursive = this.option('columnChooser.selection.recursive');
+        const column = this._columnsController.columnOption(columnIndex);
+
+        const isRecursiveWithColumns = isRecursive && column.hasColumns;
+
+        if(!isRecursiveWithColumns) {
+            column.visible ?
+                this._columnChooserList.selectItem(columnIndex) :
+                this._columnChooserList.unselectItem(columnIndex);
```
Comment (williamvinogradov) [inline js/ui/grid_core/ui.grid_core.column_chooser.js]: Minor comment.

Let's change a little formatting here:
```suggestion
            column.visible
                ? this._columnChooserList.selectItem(columnIndex)
                : this._columnChooserList.unselectItem(columnIndex);
```

File: js/ui/grid_core/ui.grid_core.column_chooser.js (hunk 2 above)
Comment (pomahtri) [reply js/ui/grid_core/ui.grid_core.column_chooser.js]: I would like to veto adding new entities before moving to ts

**Review by pomahtri — APPROVED** [review]

**Pull request merged**